htsinfer/get_library_source.py

```diff
@@ -270,26 +271,32 @@ def get_source_expression(
                 "Could not process file: {_file}"
             ) from exc
 
-        # handle case where no alignments are found
-        dat.tpm.fillna(0, inplace=True)
-
         # aggregate expression by source identifiers
-        dat[[
-            'gene_symbol',
-            'gene_id',
-            'transcript_id',
-            'short_name',
-            'taxon_id'
-        ]] = dat.target_id.str.split('|', 4, expand=True)
-        dat['source_ids'] = list(zip(dat.short_name, dat.taxon_id))
-        total_tpm = dat.tpm.sum()
-        dat_agg = dat.groupby(['source_ids'])[['tpm']].agg('sum')
-        dat_agg['source_ids'] = dat_agg.index
-        dat_agg.reset_index(drop=True, inplace=True)
-
-        # calculate percentages
-        if total_tpm != 0:
-            dat_agg.tpm = dat_agg.tpm / total_tpm * 100
-
-        # return as dictionary
-        return dat_agg.sort_values(["tpm"], ascending=False)
+        try:
+            # handle case where no alignments are found
+            dat.tpm.fillna(0, inplace=True)
+
+            dat[[
+                'gene_symbol',
+                'gene_id',
+                'transcript_id',
+                'short_name',
+                'taxon_id'
+            ]] = dat.target_id.str.split('|', 4, expand=True)
+            dat['source_ids'] = list(zip(dat.short_name, dat.taxon_id))
+            total_tpm = dat.tpm.sum()
+            dat_agg = dat.groupby(['source_ids'])[['tpm']].agg('sum')
+            dat_agg['source_ids'] = dat_agg.index
+            dat_agg.reset_index(drop=True, inplace=True)
+
+            # calculate percentages
+            if total_tpm != 0:
+                dat_agg.tpm = dat_agg.tpm / total_tpm * 100
+
+            # return as dictionary
+            return dat_agg.sort_values(["tpm"], ascending=False)
+
+        except (AttributeError, ValueError, OSError) as exc:
+            raise FileProblem(
+                "Could not process file: {_file}"
+            ) from exc
```

Comment:
Not a fan of wrapping a ton of code in some `try` - `except` block and then catching a bunch of errors. The resulting information for the user will invariably vague, just as vague as if errors weren't caught at all.

Moreover, the data is already loaded at this point, so whatever happens, it's certainly not really a `FileProblem` (in the sense that something went wrong with opening the file), but rather some sort of `ValueError` - yes, the file is there, it could be opened, but it doesn't quite look as expected.

I'd suggest the following: If you have any specific ideas of what might go wrong with a specific line or small block of code (e.g., the data frame may be empty), then catch just these problems, case by case. If not, no problem then. Maybe in the future someone will report an error/problem with the code, and then we can still catch errors accordingly.

In any case, please remove that big, catch all `try` - `except` block again.

If not,

Reply:
I agree that it's better to `try` - `catch` than add a conditional to check explicitly if the file is empty. But you should then identify the (first) statement that raises an error on an empty data frame and then only wrap that one in `try` - `catch`. But let's not just catch any kind of error that might occur but that we can't think of now. If we find more of these, raised by the same or other statements, we can still catch those errors and react accordingly. But at any given point, let's only catch those errors that we _know_ can/will happen